### PR TITLE
misc: Add OrderBy benchmark

### PR DESCRIPTION
**Pull request by jinchengchenghh**
To track the OrderBy performance and measure order by optimization performance, add this benchmark.

The key number varies with 1 to 4, benchmark result is:
```
Key Type | Dataset Size | Time(No Payload)   |  Time(With Payload)             |
---------|--------------|--------------------|-------------------------------------|
 bigint  | 0.01k        | 6.63s ~ 7.32s      | 7.50s ~ 7.95s                   |
 bigint  | 0.05k        | 7.03s ~ 7.75s      | 7.89s ~ 8.57s                   |
 bigint  | 1k           | 7.33s ~ 8.42s      | 8.33s ~ 9.44s                   |
 bigint  | 10k          | 9.83s ~ 12.26s     | 13.44s ~ 16.00s                 |
 bigint  | 100k         | 1.10ms ~ 1.60ms    | 1.89ms ~ 2.95ms                 |
 bigint  | 1000k        | 6.14ms ~ 10.07ms   | 15.48ms ~ 23.55ms               |
 varchar | 1k           | 1.89ms ~ 2.95ms    |                                 |
 varchar | 10k          | 15.48ms ~ 23.55ms  |                                 |
 varchar | 100k         | 85.17ms ~ 134.06ms |                                 |
 varchar | 1000k        | 1.31s ~ 2.06s      |                                 |
```
We can run this benchmark by `./velox_orderby_benchmark` or run single benchmark `./velox_orderby_benchmark -bm_estimate_time -bm_regex OrderBy_no-payload_1_bigint_0.01k`

From the printed metrics, we can see addInput(storeRows to RowContainer) and getOutput(extract column from RowContainer) expires most of the time, the noMoreInput(Finish time, do prefixsort) is little.
```
OrderBy_no-payload_1_bigint_0.01k                            6.52s   153.38m
Total 6.40s Input 179.60ms Output 198.20ms Finish 67.48ms
OrderBy_no-payloads_1_varchar_1k                            1.94ms    515.31
Total 2.32ms Input 261.25us Output 99.17us Finish 7.59us
```

### CODE REVIEW

**Comment by FelixYBW** [comment]
Can you add some description? like why we need it, how to run it, what's the initial performance number, etc.

File: velox/exec/SortBuffer.cpp [hunk 1]
```diff
@@ -47,9 +47,11 @@ SortBuffer::SortBuffer(
   sortedSpillColumnNames.reserve(input->size());
   sortedSpillColumnTypes.reserve(input->size());
   std::unordered_set<column_index_t> sortedChannelSet;
+  decodedInputs_.resize(input_->size());
   // Sorted key columns.
   for (column_index_t i = 0; i < sortColumnIndices.size(); ++i) {
     columnMap_.emplace_back(IdentityProjection(i, sortColumnIndices.at(i)));
+    decodedInputs_[sortColumnIndices.at(i)].second = i;
```
Comment (jinchengchenghh) [inline velox/exec/SortBuffer.cpp:54]: Since the columnMap_ contain all the columns, the index is from 0 to numColumns, we can also change it to std::vector<int>, the index is the outoutChannel, the value in the index is the inputChannel, then we could not reserve  it in decodedInputs_.
We need change function `gatherCopy`  which is only used here too.

**Comment by jinchengchenghh** [comment]
Can you help review this PR, thanks! @mbasmanova

**Comment by jinchengchenghh** [comment]
Hi, @mbasmanova Soft reminder, can you spare some time on this PR?

**Comment by mbasmanova** [comment]
@jinchengchenghh I'm traveling for work this week and on vacation next week. I have limited availability this week and none next week. I'll make sure to review this once I'm back from vacation.

In the meantime, I have some high level questions / comments.

Re-using DecodedVectors makes sense. Wondering how much improvement that brings? Is it in CPU or memory usage? 

It would be easier to review changes if they were split into 2 PR: (1) add benchmark for OrderBy; (2) optimize SortBuffer.

For the benchmark, would you describe what the benchmark measures and what is the summary of the results? The output included in the PR description is very long and hard to reason about.

@rui-mo @PHILO-HE Would you help take a first pass?

File: velox/exec/SortBuffer.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/exec/SortBuffer.cpp:54]: Are you saying that you cannot create std::vector<DecodedVector> and then call resize(n) on it? I wonder why? DecodedVector has default constructor and supports std::move. What else is missing?

File: velox/exec/SortBuffer.h [hunk 2]
```diff
@@ -94,6 +94,8 @@ class SortBuffer {
   // buffer stores the sort columns first in 'data_'.
   std::vector<IdentityProjection> columnMap_;
 
+  std::vector<std::pair<DecodedVector, column_index_t>> decodedInputs_;
```
Comment (rui-mo) [inline velox/exec/SortBuffer.h:97]: It would be better to add some comments for this variable.

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        numKeys{numKeys} {}
+
+  std::string testName;
+  size_t numRows;
+  int numVectors;
+  RowTypePtr rowType;
+  int numKeys;
+};
+
+class OrderByBenchmark {
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:49]: It appears that this benchmark and PrefixSortBenchmark share some code. I wonder if we can extract the common implementations to a separate file like SortBenchmarkUtil.cpp and reuse these methods. Thanks.

**Review by rui-mo — COMMENTED** [review]
Shall we highlight the performance improvement with this optimization in the PR description? Thanks.

**Review by rui-mo — COMMENTED** [review]
Thanks. Just one nit.

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        noPayload ? "no-payload" : "payload",
+        "bigint",
+        batchSizes,
+        rowTypes,
+        numKeys,
+        numVectors);
+  }
+
+  core::PlanNodePtr makeOrderByPlan(TestCase& test) {
+    folly::BenchmarkSuspender suspender;
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:137]: nit: seems not in-use. Could you confirm?

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    std::vector<RowVectorPtr> vectors;
+    for (auto i = 0; i < test.numVectors; ++i) {
+      vectors.emplace_back(OrderByBenchmarkUtil::fuzzRows(
+          test.rowType, test.numRows, test.numKeys, pool_.get()));
+    }
+    return makeOrderByPlan(vectors, makeOrderByKeys(test.numKeys));
+  }
+
+  core::PlanNodePtr makeOrderByPlan(
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:146]: Maybe combine the two makeOrderByPlan functions and makeOrderByKeys function because it is used only once.

**Comment by jinchengchenghh** [comment]
Do you have further comments? @rui-mo

File: velox/exec/benchmarks/OrderByBenchmarkUtil.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ */
+#pragma once
+
+#include "velox/vector/ComplexVector.h"
+
+namespace facebook::velox {
+
+class OrderByBenchmarkUtil {
+ public:
+  /// Generate bigint row types with or without Varchar.
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.h:24]: nit: Varchar -> varchar payload types

File: velox/exec/benchmarks/OrderByBenchmarkUtil.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  /// @param noPayload the return types does not include Varchar.
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.h:25]: nit: If true, the return types does not include varchar payload types.

File: velox/exec/benchmarks/OrderByBenchmarkUtil.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  /// @return row types.
+  static std::vector<RowTypePtr> bigintRowTypes(bool noPayload);
+  // Generate Varchar row types.
+  static std::vector<RowTypePtr> largeVarcharRowTypes();
+
+  /// Generate RowVector by VectorFuzzer according to rowType, for front keys
+  /// (column 0 to numKeys -2) use high
+  // nullRatio to enforce all columns to be compared.
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.h:33]: nit: add comment for numKeys as it is not obvious.

**Comment by jinchengchenghh** [comment]
Any more comments? Thanks! @rui-mo

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+
+#include "OrderByBenchmarkUtil.h"
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:19]: nit: use full include path.

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+using namespace facebook::velox;
+
+namespace {
+RowTypePtr rowWithName(std::vector<TypePtr>&& types) {
+  std::vector<std::string> names;
+  for (auto i = 0; i < types.size(); ++i) {
+    names.emplace_back(fmt::format("c{}", i));
+  }
+  return ROW(std::move(names), std::move(types));
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:28]: VectorMaker has provided a similar function 'rowType', can we use it?
https://github.com/facebookincubator/velox/blob/main/velox/vector/tests/utils/VectorMaker.cpp#L21-L28

File: velox/exec/benchmarks/OrderByBenchmarkUtil.h [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/vector/ComplexVector.h"
+
+namespace facebook::velox {
+
+class OrderByBenchmarkUtil {
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmarkUtil.h:22]: The `Util` class always contains static class like `DecimalUtil`, It's better to construct similar function to a utility class.

**Comment by jinchengchenghh** [comment]
Addressed all the comments. @rui-mo

File: velox/exec/benchmarks/PrefixSortBenchmark.cpp [hunk 12]
```diff
@@ -16,9 +16,9 @@
 #include <folly/Benchmark.h>
 #include <folly/init/Init.h>
 
+#include "OrderByBenchmarkUtil.h"
```
Comment (rui-mo) [inline velox/exec/benchmarks/PrefixSortBenchmark.cpp:19]: nit: use full include path.

File: velox/exec/benchmarks/CMakeLists.txt [hunk 13]
```diff
@@ -87,11 +87,23 @@ if(${VELOX_ENABLE_PARQUET})
...
 
 target_link_libraries(
   velox_prefixsort_benchmark
   velox_exec
   velox_vector_fuzzer
   velox_vector_test_lib
   ${FOLLY_BENCHMARK})
+
+add_executable(velox_orderby_benchmark OrderByBenchmark.cpp
+                                       OrderByBenchmarkUtil.cpp)
```
Comment (rui-mo) [inline velox/exec/benchmarks/CMakeLists.txt:101]: How about adding 'OrderByBenchmarkUtil.cpp' in a lib and link against it for velox_prefixsort_benchmark and velox_orderby_benchmark?

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+                  outputNanos += op.getOutputTiming.wallNanos;
+                }
+              }
+            }
+            uint64_t total = getCurrentTimeMicro() - start;
+            std::cout << "Total " << succinctMicros(total) << " Input "
+                      << succinctNanos(inputNanos) << " Output "
+                      << succinctNanos(outputNanos) << " Finish "
+                      << succinctNanos(finishNanos) << std::endl;
+            return 1;
```
Comment (rui-mo) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:89]: Why 1 is returned? Thanks.

Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmark.cpp:89]: The return value means run the benchmark times, other benchmark also returns 1. https://github.com/facebookincubator/velox/blob/main/velox/exec/benchmarks/FilterProjectBenchmark.cpp#L249
I generate the plan with data generated by VectorFuzzer which is an expense cost, this can help us use `./velox_orderby_benchmark -bm_estimate_time -bm_regex OrderBy_no-payload_1_bigint_0.01k` to generate data for single benchmark, instead of generating all the data even if specify one benchmark as PrefixSortBenchmark does, so I execute the plan `iteration` times.

**Comment by jinchengchenghh** [comment]
Thanks! Addressed all the comments. @rui-mo

**Review by rui-mo — APPROVED** [review]
Thanks.

**Comment by rui-mo** [comment]
> From the printed metrics, we can see addInput(storeRows to RowContainer) and getOutput(extract column from RowContainer)expires most of the time, the noMoreInput(Finish time, do prefixsort) is little.

Where can we find the metrics for addInput and getOutput, are they added in this PR?

**Comment by rui-mo** [comment]
@mbasmanova Do you have any more comments? Thanks.

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            auto plan = makeOrderByPlan(*test);
+            uint64_t inputNanos = 0;
+            uint64_t outputNanos = 0;
+            uint64_t finishNanos = 0;
+            auto start = getCurrentTimeMicro();
+            for (auto i = 0; i < iterations; ++i) {
+              std::shared_ptr<Task> task;
+              test::AssertQueryBuilder(plan).copyResults(pool_.get(), task);
+              auto stats = task->taskStats();
+              for (auto& pipeline : stats.pipelineStats) {
```
Comment (mbasmanova) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:73]: Any particular reason not to use toPlanStats function from velox/exec/PlanNodeStats.h?

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    vectors.emplace_back(OrderByBenchmarkUtil::fuzzRows(
+        test.rowType, test.numRows, test.numKeys, pool_.get()));
+
+    std::vector<std::string> keys;
+    keys.reserve(test.numKeys);
+    for (auto i = 0; i < test.numKeys; i++) {
+      keys.emplace_back(fmt::format("c{} ASC NULLS LAST", i));
+    }
+
+    auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
```
Comment (mbasmanova) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:108]: PlanNodeIdGenerator is not needed for straight-line plans.

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    core::PlanNodeId orderNodeId;
+    return test::PlanBuilder(planNodeIdGenerator)
+        .values(vectors)
+        .orderBy(keys, false)
+        .capturePlanNodeId(orderNodeId)
```
Comment (mbasmanova) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:113]: This is not used.

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+#include "velox/exec/tests/utils/AssertQueryBuilder.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+
+namespace {
+
+struct TestCase {
+  explicit TestCase(
```
Comment (mbasmanova) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:31]: This constructor is not needed.

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+
+    std::vector<std::string> keys;
+    keys.reserve(test.numKeys);
+    for (auto i = 0; i < test.numKeys; i++) {
+      keys.emplace_back(fmt::format("c{} ASC NULLS LAST", i));
+    }
+
+    auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+    core::PlanNodeId orderNodeId;
+    return test::PlanBuilder(planNodeIdGenerator)
```
Comment (mbasmanova) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:110]: This plan runs orderBy and then copies results. The cost of copy is probably significant, hence, I suggest to add a node that discards the results, i.e. count(1) aggregation or limit(1) or similar. Alternatively, copyResults can be replaced with a new method 'countResults' that simply counts number of rows produced by the query. 

Furthermore, it might be better to use Task::next API  to run query single-threaded. This allows to measure the cost of OrderBy more accurately.

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+  // Fuzz keys: for front keys (column 0 to numKeys -2) use high
+  // nullRatio to enforce all columns to be compared.
+  {
+    for (auto i = 0; i < numKeys - 1; ++i) {
+      children.push_back(fuzzerWithNulls.fuzz(rowType->childAt(i)));
+    }
+    children.push_back(fuzzer.fuzz(rowType->childAt(numKeys - 1)));
+  }
+  // Fuzz payload.
+  {
```
Comment (mbasmanova) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:187]: this extra scope is not needed

**Review by mbasmanova — COMMENTED** [review]
@jinchengchenghh Thank you for adding this benchmark. Some comments.

Can you summarize current benchmark results and add that summary to the PR description?

File: velox/exec/benchmarks/OrderByBenchmark.cpp (hunk 15 above)
Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmark.cpp:73]: PlanNodeStats combine them in one metric `cpuWallTiming`. https://github.com/facebookincubator/velox/blob/main/velox/exec/PlanNodeStats.h#L82
If you think we could collect each of them, I can help to refactor it.

File: velox/exec/benchmarks/OrderByBenchmark.cpp (hunk 19 above)
Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmark.cpp:110]: Thanks for catch this. Changed.

File: velox/exec/benchmarks/OrderByBenchmark.cpp (hunk 18 above)
Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmark.cpp]: It is used for std::make_unique<TestCase>, I have refactored it to drop the constructor.

File: velox/exec/benchmarks/OrderByBenchmark.cpp (hunk 15 above)
Comment (mbasmanova) [reply velox/exec/benchmarks/OrderByBenchmark.cpp:58]: Sure, if it is helpful to keep the breakdown, let's update toPlanStats to produce both total CPU and input/output/etc. CPU.

File: velox/exec/tests/utils/AssertQueryBuilder.h [hunk 21]
```diff
@@ -175,6 +175,9 @@ class AssertQueryBuilder {
       memory::MemoryPool* pool,
       std::shared_ptr<Task>& task);
 
+  /// Only run the query, no check for the result.
+  std::shared_ptr<Task> runQuery();
```
Comment (mbasmanova) [inline velox/exec/tests/utils/AssertQueryBuilder.h:179]: 'Query' is not necessary since it is a method in XxxQueryXxx class.

I wonder if 'run' could be confusing as it is not clear whether the method just starts running and expects the caller to monitor / wait for completion or the method actually waits for completion.

Perhaps, runToCompletion() or countResults() is clearer.

```
std::shared_ptr<Task> runToCompletion();

int64_t countResults(std::shared_ptr<Task>& task);
```

File: velox/exec/tests/utils/AssertQueryBuilder.h [hunk 22]
```diff
@@ -175,6 +175,9 @@ class AssertQueryBuilder {
       memory::MemoryPool* pool,
       std::shared_ptr<Task>& task);
 
+  /// Run the query and return the result number.
+  vector_size_t AssertQueryBuilder::countResults(std::shared_ptr<Task>& task);
```
Comment (mbasmanova) [inline velox/exec/tests/utils/AssertQueryBuilder.h:179]: This needs to be int64_t.

vector_size_t is int32_t which is enough for vector size, but not enough for the sum across many vectors.

File: velox/exec/tests/utils/AssertQueryBuilder.h [hunk 23]
```diff
@@ -175,6 +175,9 @@ class AssertQueryBuilder {
       memory::MemoryPool* pool,
       std::shared_ptr<Task>& task);
 
+  /// Run the query and return the result number.
```
Comment (mbasmanova) [inline velox/exec/tests/utils/AssertQueryBuilder.h:178]: > the result number

the number of result rows.

File: velox/exec/benchmarks/OrderByBenchmark.cpp (hunk 15 above)
Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmark.cpp:59]: Added in https://github.com/facebookincubator/velox/pull/10986, thanks!

**Comment by jinchengchenghh** [comment]
Can you help review again? Thanks! @mbasmanova 
Refactor task stats after https://github.com/facebookincubator/velox/pull/10986 merged.

**Comment by jinchengchenghh** [comment]
Resolved all the comments and pass CI, could you help review again? Thanks! @mbasmanova

File: velox/exec/tests/utils/AssertQueryBuilder.h [hunk 24]
```diff
@@ -183,6 +183,9 @@ class AssertQueryBuilder {
       memory::MemoryPool* pool,
       std::shared_ptr<Task>& task);
 
+  /// Run the query and return the number of result rows.
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/AssertQueryBuilder.h:186]: Is this to optimize the benchmark execution or? Thanks!

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+namespace {
+
+void addBenchmark(
+    const std::string& prefix,
+    const std::string& keyName,
+    const std::vector<vector_size_t>& batchSizes,
+    const std::vector<RowTypePtr>& rowTypes,
+    const std::vector<int>& numKeys,
+    int32_t iterations,
+    std::function<void(
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:35]: Can you define an alias for benchmark callback? Thanks!

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 26]
```diff
@@ -0,0 +1,308 @@
...
+        const std::string& testName,
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:36]: testName is not used?

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 27]
```diff
@@ -0,0 +1,308 @@
...
+        size_t numRows,
+        const RowTypePtr& rowType,
+        int iterations,
+        int numKeys)>& benchmark) {
+  for (auto batchSize : batchSizes) {
+    for (auto i = 0; i < rowTypes.size(); ++i) {
+      const auto name = fmt::format(
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:43]: testName or benchmarkName? Thanks!

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 28]
```diff
@@ -0,0 +1,308 @@
...
+          "{}_{}_{}_{}k", prefix, numKeys[i], keyName, batchSize / 1000.0);
+      benchmark(name, batchSize, rowTypes[i], iterations, numKeys[i]);
+    }
+  }
+}
+
+std::vector<RowTypePtr> smallintRowTypes(bool noPayload) {
+  if (noPayload) {
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:51]: nit: use
```
if () {
  return;
}
...
```

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 29]
```diff
@@ -0,0 +1,308 @@
...
+    bool noPayload,
+    int numIterations,
+    const std::vector<vector_size_t>& batchSizes,
+    std::function<void(
+        const std::string& testName,
+        size_t numRows,
+        const RowTypePtr& rowType,
+        int iterations,
+        int numKeys)> benchmark) {
+  std::vector<RowTypePtr> rowTypes = bigintRowTypes(noPayload);
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:113]: consts

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 30]
```diff
@@ -0,0 +1,308 @@
...
+    bool noPayload,
+    int numIterations,
+    const std::vector<vector_size_t>& batchSizes,
+    std::function<void(
+        const std::string& testName,
+        size_t numRows,
+        const RowTypePtr& rowType,
+        int iterations,
+        int numKeys)> benchmark) {
+  std::vector<RowTypePtr> rowTypes = smallintRowTypes(noPayload);
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:204]: dittos

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 31]
```diff
@@ -0,0 +1,308 @@
...
+  VectorFuzzer fuzzer({.vectorSize = static_cast<size_t>(numRows)}, pool);
+  VectorFuzzer fuzzerWithNulls(
+      {.vectorSize = static_cast<size_t>(numRows),
+       .nullRatio = FLAGS_data_null_ratio},
+      pool);
+  std::vector<VectorPtr> children;
+
+  // Fuzz keys: for front keys (column 0 to numKeys -2) use high
+  // nullRatio to enforce all columns to be compared.
+  for (auto i = 0; i < numKeys - 1; ++i) {
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:279]: Does this reflect the actual use scenarios? Can we simplify the case with null and not-null to ease analysis? Thanks!

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 32]
```diff
@@ -0,0 +1,120 @@
...
+      int numKeys) {
+    TestCase testCase = {numRows, rowType, numKeys};
+    {
+      folly::addBenchmark(
+          __FILE__,
+          "OrderBy_" + testName,
+          [test = testCase, iterations = std::max(1, iterations / 10), this]() {
+            core::PlanNodeId orderByNodeId;
+            auto plan = makeOrderByPlan(test, orderByNodeId);
+            uint64_t inputNanos = 0;
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:53]: nit: s/inputNanos/inputNs/

File: velox/exec/benchmarks/OrderByBenchmark.cpp [hunk 33]
```diff
@@ -0,0 +1,120 @@
...
+            for (auto i = 0; i < iterations; ++i) {
+              std::shared_ptr<Task> task;
+              test::AssertQueryBuilder(plan).countResults(task);
+              auto taskStats = exec::toPlanStats(task->taskStats());
+              auto& stats = taskStats.at(orderByNodeId);
+              inputNanos += stats.addInputTiming.wallNanos;
+              finishNanos += stats.finishTiming.wallNanos;
+              outputNanos += stats.getOutputTiming.wallNanos;
+            }
+            uint64_t total = getCurrentTimeMicro() - start;
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmark.cpp:66]: consts

**Review by xiaoxmeng — COMMENTED** [review]
@jinchengchenghh overall looks good % nits. Thanks!

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp (hunk 31 above)
Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:279]: Actual cases columns maybe null or not null, so we need both of them. The default null ratio 0.7 is not actual use scenarios, it is just used to compute all the columns in sort, we can change it by FLAGS_data_null_ratio

File: velox/exec/tests/utils/AssertQueryBuilder.h (hunk 24 above)
Comment (jinchengchenghh) [reply velox/exec/tests/utils/AssertQueryBuilder.h:186]: Yes, it is used to optimize the benchmark execution. Before that, we only have the API to copy the result, which may interference the benchmark performance.

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp (hunk 26 above)
Comment (jinchengchenghh) [reply velox/exec/benchmarks/OrderByBenchmarkUtil.cpp]: It is used here https://github.com/facebookincubator/velox/pull/10041/files#diff-a3c00ce2041a5d08e67ea881239652bfcc9eab6eea746efc1391810332287c29R49

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp [hunk 34]
```diff
@@ -0,0 +1,248 @@
...
+    int numKeys)>;
+
+void addBenchmark(
+    const std::string& prefix,
+    const std::string& keyName,
+    const std::vector<vector_size_t>& batchSizes,
+    const std::vector<RowTypePtr>& rowTypes,
+    const std::vector<int>& numKeys,
+    int32_t iterations,
+    const OrderByBenchmarkFunction& benchmark) {
```
Comment (xiaoxmeng) [inline velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:42]:
```
s/benchmark/benchmarkFunc/
```

File: velox/exec/benchmarks/OrderByBenchmarkUtil.cpp (hunk 31 above)
Comment (xiaoxmeng) [reply velox/exec/benchmarks/OrderByBenchmarkUtil.cpp:223]: Then why not have all the key columns with null ratio set to keep it simple?

File: velox/exec/tests/utils/AssertQueryBuilder.h (hunk 24 above)
Comment (xiaoxmeng) [reply velox/exec/tests/utils/AssertQueryBuilder.h:186]: It seems that you use something similar to velox/tool/trace/TraceReplayTaskRunner.h
Maybe rename countResults to runWithoutResults?

**Review by xiaoxmeng — COMMENTED** [review]
@jinchengchenghh thanks for the update!

**Review by xiaoxmeng — APPROVED** [review]
